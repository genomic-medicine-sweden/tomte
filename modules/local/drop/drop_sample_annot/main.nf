process DROP_SAMPLE_ANNOT {
    tag "DROP_annot_file"
    label 'process_low'

    container "docker.io/clinicalgenomics/drop:1.4.0"

    input:
    tuple val(meta2), path(gtf)
    tuple val(ids), val(single_ends), val(strandednesses), val(sex), path(bam), path(bai)
    path(ref_gene_counts)
    path(ref_annot)
    val(drop_group_samples_ae)
    val(drop_group_samples_as)

    output:
    path('sample_annotation.tsv'), emit: drop_annot
    path "versions.yml"   , emit: versions

    when:
    // Exit if running this module with -profile conda / -profile mamba
    workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() == 0 ||
        { log.error("Local DROP module does not support Conda. Please use Docker / Singularity / Podman instead."); return false }

    script:
    def id = ids.join(' ')
    def single_end = single_ends.join(' ')
    def sex_drop = sex.collect { it.replace("1","M").replace("2","F").replace("0","NA").replace("other","NA") }.join(' ')
    def strandedness = strandednesses.join(' ')
    def drop_group = "${drop_group_samples_ae},${drop_group_samples_as}".replace(" ","").replace("[","").replace("]","")
    def reference_count_file = ref_gene_counts ? "--ref_count_file ${ref_gene_counts}" : ''
    def reference_annotation = ref_annot ? "--ref_annot ${ref_annot}" : ''
    """
    # Single_end values are only provided if you start from fastq
    SINGLE_ENDS=(${single_end})
    BAMS=(${bam.join(' ')})

    # Check if single_end values are provided
    updated_single_ends=()
    for ((i=0; i<\${#SINGLE_ENDS[@]}; i++)); do
        if [[ "\${SINGLE_ENDS[i]}" == "null" ]]; then
            result=\$(samtools view -c -f 1 "\${BAMS[i]}" | awk '{print \$1 == 0 ? "true" : "false"}')
            updated_single_ends+=("\$result")
        else
            updated_single_ends+=("\${SINGLE_ENDS[i]}")
        fi
    done

    # Convert updated_single_ends array to space-separated string and save to file
    echo "\${updated_single_ends[*]}" > updated_single_ends.txt

    drop_sample_annot.py \\
        --bam ${bam.join(' ')} \\
        --samples $id \\
        --strandedness $strandedness \\
        --single_end \$(cat updated_single_ends.txt) \\
        --sex $sex_drop \\
        $reference_count_file \\
        $reference_annotation \\
        --drop_group_sample $drop_group \\
        --gtf ${gtf} \\
        --output sample_annotation.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        drop_sample_annot: \$(drop_sample_annot.py --version)
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """

    stub:
    """
    touch sample_annotation.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        drop_sample_annot: \$(drop_sample_annot.py --version)
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """
}
