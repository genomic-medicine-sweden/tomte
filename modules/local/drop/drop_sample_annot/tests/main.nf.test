nextflow_process {

    name "Test Process DROP_SAMPLE_ANNOT"
    script "../main.nf"
    process "DROP_SAMPLE_ANNOT"
    config "./nextflow.config"
    tag "modules"
    tag "modules_local"
    tag "drop"
    tag "drop_sample_annot"

    setup {
        run("GUNZIP", alias: "GUNZIP_FASTA") {
            script "../../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'fasta' ],
                    file("${projectDir}/test_data/grch37_chr21.fa.gz", checkIfExists: true)
                ])
                """
            }
        }
    }

    test("Should run without failures") {

        when {
            process {
                """
                input[0] = GUNZIP_FASTA.out.gunzip
                input[1] = Channel.of([
                            ['ACC5963A1'], [false], ['reverse'], ['NA'], ["${projectDir}/test_data/test_ACC5963A1.vcf.gz"], ["${projectDir}/test_data/test_ACC5963A1.vcf.gz.tbi"], ['ACC5963A1'],
                            file("${projectDir}/test_data/ACC5963A1.bam", checkIfExists: true),
                            file("${projectDir}/test_data/ACC5963A1.bam.bai", checkIfExists: true)
                        ])
                input[2] = channel.fromPath("${projectDir}/test_data/drop_data/geneCounts.tsv.gz", checkIfExists: true)
                input[3] = channel.fromPath("${projectDir}/test_data/drop_data/sampleAnnotation.tsv", checkIfExists: true)
                input[4] = 'outrider'
                input[5] = 'fraser'
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }

    }

}
