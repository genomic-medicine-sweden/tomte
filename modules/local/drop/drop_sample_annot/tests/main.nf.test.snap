{
    "Should run without failures": {
        "content": [
            {
                "0": [
                    "sample_annotation.tsv:md5,b4280c567590c13b784215af25355a82"
                ],
                "1": [
                    "versions.yml:md5,1d7db4fd1f42c30efe32207042fbc3bd"
                ],
                "drop_annot": [
                    "sample_annotation.tsv:md5,b4280c567590c13b784215af25355a82"
                ],
                "versions": [
                    "versions.yml:md5,1d7db4fd1f42c30efe32207042fbc3bd"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.5"
        },
        "timestamp": "2025-08-12T21:47:14.75954614"
    }
}